### PR TITLE
Fix recursive update in AsyncConnectionProvider

### LINKED ISSUE #3691: Recursive call in AsyncConnectionProvider
## Bug Report

We have stumbled on a recursive update error in the AsyncConnectionProvider when using Redis to back a Spring Session used in Spring Security.

#### Current Behavior

When a cold instance of our app receives new requests with authenticated session cookies, the session lookup lands in AsyncConnectionProvider.getSynchronizer via the PooledClusterConnectionProvider. 

As this is a freshly started app instance the connections map in AsyncConnectionProvider is empty, so we call into the connections.computeIfAbsent mapping function. That mapping function ultimately leads to a Mono.subscribe that ends up invoking SecurityContextHolder.getContext() which then triggers another session lookup (while the mapping function is still in flight).

This leads to a recursive update exception being thrown from ConcurrentHashMap.putIfAbsent.

<details>
<summary>Stack trace</summary>

```java
 - Reactor/context bridge:
 ```text
   io.lettuce.core.AbstractRedisClient.initializeChannelAsync(AbstractRedisClient.java:411)
   -> reactor.core.publisher.Mono.subscribe(Mono.java:4450)
   -> reactor.core.publisher.Mono.subscribe(Mono.java:4478)
   -> reactor.core.publisher.Mono.subscribe(Mono.java:4542)
   -> reactor.core.publisher.Mono.subscribeWith(Mono.java:4641)
   -> reactor.core.publisher.Mono.subscribe(Mono.java:4576)
   -> reactor.core.publisher.MonoCallable.subscribe(MonoCallable.java:48)
   -> reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.onSubscribe(MonoPeekTerminal.java:152)
   -> reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onSubscribe(FluxPeekFuseable.java:178)
   -> org.springframework.security.config.annotation.web.configuration.SecurityReactorContextConfiguration$SecurityReactorContextSubscriber.onSubscribe(SecurityReactorContextConfiguration.java:181)
   -> reactor.core.publisher.FluxContextWriteRestoringThreadLocalsFuseable$FuseableContextWriteRestoringThreadLocalsSubscriber.onSubscribe(FluxContextWriteRestoringThreadLocalsFuseable.java:102)
   -> reactor.core.publisher.ContextPropagation.setThreadLocals(ContextPropagation.java:84)
   -> io.micrometer.context.DefaultContextSnapshotFactory.setThreadLocalsFrom(DefaultContextSnapshotFactory.java:109)
   -> io.micrometer.context.DefaultContextSnapshotFactory.setAllThreadLocalsFrom(DefaultContextSnapshotFactory.java:128)
   -> io.micrometer.context.DefaultContextSnapshot.clearThreadLocal(DefaultContextSnapshot.java:110)
   -> org.springframework.security.core.context.SecurityContextHolderThreadLocalAccessor.getValue(SecurityContextHolderThreadLocalAccessor.java:37)
   -> org.springframework.security.core.context.SecurityContextHolderThreadLocalAccessor.getValue(SecurityContextHolderThreadLocalAccessor.java:46)
   -> org.springframework.security.core.context.SecurityContextHolder.getContext(SecurityContextHolder.java:125)
 ```

 - Spring Security / Spring Session re-entry:
 ```text
   -> org.springframework.security.core.context.ThreadLocalSecurityContextHolderStrategy.getContext(ThreadLocalSecurityContextHolderStrategy.java:43)
   -> org.springframework.security.core.context.ThreadLocalSecurityContextHolderStrategy.lambda$setDeferredContext$2(ThreadLocalSecurityContextHolderStrategy.java:67)
   -> org.springframework.security.web.context.DelegatingSecurityContextRepository$DelegatingDeferredSecurityContext.get(DelegatingSecurityContextRepository.java:96)
   -> org.springframework.security.web.context.DelegatingSecurityContextRepository$DelegatingDeferredSecurityContext.get(DelegatingSecurityContextRepository.java:109)
   -> org.springframework.security.web.context.SupplierDeferredSecurityContext.get(SupplierDeferredSecurityContext.java:33)
   -> org.springframework.security.web.context.SupplierDeferredSecurityContext.get(SupplierDeferredSecurityContext.java:52)
   -> org.springframework.security.web.context.SupplierDeferredSecurityContext.init(SupplierDeferredSecurityContext.java:67)
   -> org.springframework.security.web.context.HttpSessionSecurityContextRepository.lambda$loadDeferredContext$0(HttpSessionSecurityContextRepository.java:145)
   -> jakarta.servlet.http.HttpServletRequestWrapper.getSession(HttpServletRequestWrapper.java:221)
   -> org.springframework.session.web.http.SessionRepositoryFilter$SessionRepositoryRequestWrapper.getSession(SessionRepositoryFilter.java:193)
   -> org.springframework.session.web.http.SessionRepositoryFilter$SessionRepositoryRequestWrapper.getSession(SessionRepositoryFilter.java:286)
   -> org.springframework.session.web.http.SessionRepositoryFilter$SessionRepositoryRequestWrapper.getRequestedSession(SessionRepositoryFilter.java:352)
   -> org.springframework.session.data.redis.RedisSessionRepository.findById(RedisSessionRepository.java:45)
   -> org.springframework.session.data.redis.RedisSessionRepository.findById(RedisSessionRepository.java:138)
   -> org.springframework.data.redis.core.DefaultHashOperations.entries(DefaultHashOperations.java:323)
   -> org.springframework.data.redis.connection.lettuce.LettuceHashCommands.hGetAll(LettuceHashCommands.java:110)
   -> io.lettuce.core.AbstractRedisAsyncCommands.hgetall(AbstractRedisAsyncCommands.java:1272)
   -> io.lettuce.core.cluster.StatefulRedisClusterConnectionImpl.dispatch(StatefulRedisClusterConnectionImpl.java:242)
   -> io.lettuce.core.cluster.ClusterDistributionChannelWriter.doWrite(ClusterDistributionChannelWriter.java:270)
 ```

 - Representative outermost-to-innermost summary:
 ```text
   Jetty request thread
   -> Spring Security filter chain
   -> RememberMeAuthenticationFilter
   -> SecurityContextHolder.getContext()
   -> HttpSessionSecurityContextRepository.loadDeferredContext(...)
   -> request.getSession(false)
   -> SessionRepositoryRequestWrapper.getRequestedSession()
   -> RedisSessionRepository.findById(sessionId)
   -> DefaultHashOperations.entries(sessionKey)
   -> LettuceHashCommands.hGetAll(sessionKey)
   -> StatefulRedisClusterConnectionImpl.dispatch(...)
   -> ClusterDistributionChannelWriter.doWrite(...)
   -> PooledClusterConnectionProvider.getConnectionAsync(slot/key)
   -> AsyncConnectionProvider.getSynchronizer(outer key)
   -> ConcurrentHashMap.computeIfAbsent(outer key, mapping)
   -> DefaultClusterNodeConnectionFactory.apply(...)
   -> RedisClusterClient.connectToNodeAsync(...)
   -> RedisClusterClient.connectStatefulAsync(...)
   -> AbstractRedisClient.initializeChannelAsync(...)
   -> Reactor Mono.subscribe(...)
   -> Reactor / Micrometer / Spring Security context restoration
   -> SecurityContextHolder.getContext()
   -> SessionRepositoryRequestWrapper.getRequestedSession()
   -> RedisSessionRepository.findById(sessionId)
   -> LettuceHashCommands.hGetAll(sessionKey)
   -> ClusterDistributionChannelWriter.doWrite(...)
   -> PooledClusterConnectionProvider.getConnectionAsync(slot/key)
   -> AsyncConnectionProvider.getSynchronizer(same key)
   -> ConcurrentHashMap.computeIfAbsent(same key, ...)
   -> IllegalStateException: Recursive update

```

</details>

#### Input Code

<details>
<summary>Input Code</summary>

```java
// your code here;
```

</details>

#### Expected behavior/code

No recursive update to happen.

#### Environment

- Lettuce version(s): 6.6.0.RELEASE
- Redis version: N/A

#### Possible Solution

I have a draft PR that effectively changes the `putIfAbsent` to take a placeholder `Sync`. That `Sync` triggers the existing `connectionFactoy.apply` that was happening inside the `putIfAbsent` mapping function but now does it outside it, meaning that any recursive calls to `putIfAbsent` just receive the placeholder `Sync` until it completes, at which point the connection is updated on the `Sync` object

#### Additional context

The recursive loop was also found in the `RedisClusterClient.connectToNodeAsync` method when the connectionFuture throws an exception, as this calls `connection.closeAsync()` which loops back to the channel writer and ultimately back to the `putIfAbsent` call. We didn't hit this in our case but found that path whilst trying to trace what was going on

### PR DESCRIPTION
**Pull request by pauldthomson**
Make sure that:

- [x] You have read the [contribution guidelines](https://github.com/lettuce-io/lettuce-core/blob/main/.github/CONTRIBUTING.md).
- [x] You have created a feature request first to discuss your contribution intent. Please reference the feature request ticket number in the pull request.
- [x] You applied code formatting rules using the `mvn formatter:format` target. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.

 

Fixes: #3691 

Prevent recursive update error being thrown during connection establishment when using this library via Spring Session. See the linked issue for more details. 

---

> [!NOTE]
> **Medium Risk**
> Touches core connection acquisition/closing and concurrency semantics used by cluster and master/replica modes; changes are well-covered by new unit/integration tests but could affect connection lifecycle edge cases under load.
> 
> **Overview**
> Fixes connection acquisition races/reentrancy by refactoring `AsyncConnectionProvider` to insert a per-key placeholder first and start the actual connect attempt asynchronously via a provided `Executor`, avoiding `computeIfAbsent`/recursive map update failures.
> 
> The provider API is simplified to always return `CompletableFuture` waiters so caller cancellation no longer cancels the shared underlying connection attempt; close behavior is updated to properly cancel/await pending connects and remove failed placeholders. Cluster and master/replica connection providers are updated to use the new provider API/executor and to preserve remote address information for connection errors.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 95792bd6d436f8dee546673a9ec43b8dd9110436. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/main/java/io/lettuce/core/internal/AsyncConnectionProvider.java [hunk 1]
```diff
@@ -211,58 +241,168 @@ static class Sync<K, T extends AsyncCloseable, F extends CompletionStage<T>> {
...
+        public boolean isCloseRequested() {
+            return closeRequested.get();
         }
 
-        public F getConnection() {
-            return future;
+        public void completeCloseWithoutConnection() {
+            if (closeRequested.get()) {
+                closeFuture.complete(null);
+            }
```
[the remark this answers is not in the document]
Comment (pauldthomson) [reply src/main/java/io/lettuce/core/internal/AsyncConnectionProvider.java:397]: But there's a `if (closed)` before we ever call `startConnect` so we can't get into `startConnect` when `closed` is true. That same block also sets `closeRequested`.

[the remark this answers is not in the document]
Comment (pauldthomson) [reply src/main/java/io/lettuce/core/internal/AsyncConnectionProvider.java:397]: cursor review

[the remark this answers is not in the document]
Comment (pauldthomson) [reply src/main/java/io/lettuce/core/internal/AsyncConnectionProvider.java:397]: OK I don't know if there's a way to "prompt" `bugbot` with feedback but I believe the above comment is wrong?

[the remark this answers is not in the document]
Comment (pauldthomson) [reply src/main/java/io/lettuce/core/internal/AsyncConnectionProvider.java:397]: OK I see it now....

[the remark this answers is not in the document]
Comment (pauldthomson) [reply src/main/java/io/lettuce/core/internal/AsyncConnectionProvider.java:397]: @cursor review

**Comment by a-TODO-rov** [comment]
Thanks @pauldthomson !
I am on it !

**Comment by pauldthomson** [comment]
@a-TODO-rov 🍻 

Note the repro I sent was using `6.6.0.RELEASE` (and the Spring Boot/Session version it comes via) but also worked with this version. We will (ideally) need this cherry-picked onto that too but not sure what your patching policy is for older releases.. Let me know and we can figure out what to do.

**Comment by a-TODO-rov** [comment]
check my [reply](https://github.com/redis/lettuce/issues/3691#issuecomment-4297160593) in the issue
